# Практическое занятие 13. ПРИНЯТИЕ РЕШЕНИЙ В УСЛОВИЯХ РИСКА

In [1]:
import numpy as np

# Данные из задания
purchase_volumes = [500, 1000, 2000]  # Объем закупок (фунты)
probabilities = [0.2, 0.4, 0.4]  # Вероятности соответствующих объемов заказов
buy_price = 1.5  # Стоимость закупки за фунт
sell_price = 2.5  # Стоимость продажи за фунт
discount_price = 0.5  # Стоимость продажи непроданных моллюсков за фунт

# Функция для расчета прибыли
def calculate_profit_or_loss(purchase_volume):
    profits = []
    for demand, probability in zip(purchase_volumes, probabilities):
        # Расчет прибыли/убытка при различных уровнях спроса
        sold_amount = min(purchase_volume, demand)  # Количество проданных моллюсков
        unsold_amount = purchase_volume - sold_amount  # Количество непроданных моллюсков
        profit = (sold_amount * sell_price + unsold_amount * discount_price) - purchase_volume * buy_price
        profits.append(profit * probability)
    return sum(profits)

# Расчет ожидаемой прибыли или убытков для каждого объема закупок
expected_profits = {volume: calculate_profit_or_loss(volume) for volume in purchase_volumes}

# Определение оптимального объема закупок
optimal_purchase_volume = max(expected_profits, key=expected_profits.get)

# Расчет стандартного отклонения доходности
def calculate_std(purchase_volume):
    profit_variance = []
    expected_profit = calculate_profit_or_loss(purchase_volume)
    for demand, probability in zip(purchase_volumes, probabilities):
        sold_amount = min(purchase_volume, demand)
        unsold_amount = purchase_volume - sold_amount
        profit = (sold_amount * sell_price + unsold_amount * discount_price) - purchase_volume * buy_price
        profit_variance.append(probability * (profit - expected_profit) ** 2)
    return np.sqrt(sum(profit_variance))

std_profits = {volume: calculate_std(volume) for volume in purchase_volumes}

expected_profits, optimal_purchase_volume, std_profits

({500: 500.0, 1000: 800.0, 2000: 600.0},
 1000,
 {500: 0.0, 1000: 400.0, 2000: 1200.0})

1.  **Размер ожидаемой прибыли или убытков для каждого объема закупок**
    
    *   При закупке 500 фунтов ожидаемая прибыль составляет $500.
    *   При закупке 1000 фунтов ожидаемая прибыль составляет $800.
    *   При закупке 2000 фунтов ожидаемая прибыль составляет $600.
2.  **Оптимальный объем закупок моллюсков**
    
    *   Исходя из вычисленной ожидаемой прибыли, оптимальным объемом закупок является 1000 фунтов, так как при этом объеме ожидаемая прибыль максимальна.
3.  **Стандартное отклонение доходности при закупке**
    
    * При закупке 500 фунтов стандартное отклонение составляет $0 (нет риска).
    *   При закупке 1000 фунтов стандартное отклонение составляет $400.
    *   При закупке 2000 фунтов стандартное отклонение составляет $1200.



In [2]:
# Вычисление ожидаемых размеров упущенной выгоды
def calculate_opportunity_loss(purchase_volume):
    losses = []
    max_profit = max(expected_profits.values())
    expected_profit = calculate_profit_or_loss(purchase_volume)
    opportunity_loss = max_profit - expected_profit
    return opportunity_loss

opportunity_losses = {volume: calculate_opportunity_loss(volume) for volume in purchase_volumes}

# Нахождение коэффициента вариации доходности
def calculate_coefficient_of_variation(purchase_volume):
    std_dev = calculate_std(purchase_volume)
    mean_profit = calculate_profit_or_loss(purchase_volume)
    if mean_profit == 0:
        return float('inf')  # Для избежания деления на ноль
    return std_dev / mean_profit

coefficients_of_variation = {volume: calculate_coefficient_of_variation(volume) for volume in purchase_volumes}

# Вычисление отношения «доходность/риск»
profit_risk_ratio = {volume: expected_profits[volume] / std_profits[volume] if std_profits[volume] != 0 else float('inf') for volume in purchase_volumes}

# Выбор оптимального размера закупок, руководствуясь ответами на вопросы 2 и 4
# Используется критерий максимальной ожидаемой прибыли и минимальной упущенной выгоды
optimal_volume_by_profit_and_loss = min(opportunity_losses, key=opportunity_losses.get)

opportunity_losses, coefficients_of_variation, profit_risk_ratio, optimal_volume_by_profit_and_loss

({500: 300.0, 1000: 0.0, 2000: 200.0},
 {500: 0.0, 1000: 0.5, 2000: 2.0},
 {500: inf, 1000: 2.0, 2000: 0.5},
 1000)

4.   **Ожидаемые размеры упущенной выгоды**:
    
    *   При закупке 500 фунтов: упущенная выгода составляет $300.
    *   При закупке 1000 фунтов: упущенная выгода равна $0.
    *   При закупке 2000 фунтов: упущенная выгода составляет $200.
5.   **Коэффициент вариации доходности**:
    
    *   При закупке 500 фунтов: коэффициент вариации составляет 0 (нет риска).
    *   При закупке 1000 фунтов: коэффициент вариации составляет 0.5.
    *   При закупке 2000 фунтов: коэффициент вариации составляет 2.
6.  **Отношение «доходность/риск»**:
    
    *   При закупке 500 фунтов: отношение равно бесконечности (нет риска).
    *   При закупке 1000 фунтов: отношение составляет 2.
    *   При закупке 2000 фунтов: отношение составляет 0.5.
7.   **Оптимальный размер закупок**:
    
    *   На основе ожидаемой прибыли и упущенной выгоды, оптимальным объемом закупок является 1000 фунтов, так как при этом объеме ожидаемая прибыль максимальна, а упущенная выгода минимальна.

In [3]:
# Пункт 8 - Сравнение результатов решения задач 2, 4, 5 и 6

# Для сравнения, мы соберем результаты по каждому критерию для всех объемов закупок
comparison_results = {
    volume: {
        "Expected Profit": expected_profits[volume],
        "Opportunity Loss": opportunity_losses[volume],
        "Coefficient of Variation": coefficients_of_variation[volume],
        "Profit/Risk Ratio": profit_risk_ratio[volume]
    }
    for volume in purchase_volumes
}

comparison_results

{500: {'Expected Profit': 500.0,
  'Opportunity Loss': 300.0,
  'Coefficient of Variation': 0.0,
  'Profit/Risk Ratio': inf},
 1000: {'Expected Profit': 800.0,
  'Opportunity Loss': 0.0,
  'Coefficient of Variation': 0.5,
  'Profit/Risk Ratio': 2.0},
 2000: {'Expected Profit': 600.0,
  'Opportunity Loss': 200.0,
  'Coefficient of Variation': 2.0,
  'Profit/Risk Ratio': 0.5}}

8. Вот результаты сравнения для задач 2, 4, 5 и 6:

*   **При закупке 500 фунтов**:
    
    *   Ожидаемая прибыль: $500.
    *   Упущенная выгода: $300.
    *   Коэффициент вариации: 0 (нет риска).
    *   Отношение «доходность/риск»: бесконечность (нет риска).
*   **При закупке 1000 фунтов**:
    
    *   Ожидаемая прибыль: $800.
    *   Упущенная выгода: $0.
    *   Коэффициент вариации: 0.5.
    *   Отношение «доходность/риск»: 2.
*   **При закупке 2000 фунтов**:
    
    *   Ожидаемая прибыль: $600.
    *   Упущенная выгода: $200.
    *   Коэффициент вариации: 2.
    *   Отношение «доходность/риск»: 0.5.

**Анализ сравнения**:

*   При закупке 500 фунтов, нет риска (коэффициент вариации равен нулю), но ожидаемая прибыль ниже и упущенная выгода выше, чем при других объемах закупок.
*   При закупке 1000 фунтов, достигается максимальная ожидаемая прибыль с минимальной упущенной выгодой и приемлемым риском (отношение «доходность/риск» равно 2).
*   При закупке 2000 фунтов, хотя ожидаемая прибыль выше, чем при 500 фунтах, но ниже, чем при 1000 фунтах, и сопровождается самым высоким риском (коэффициент вариации равен 2).

In [4]:
# Пункт 9 - Решение задач 1-7 с измененной стоимостью продажи (3 долл. за фунт)

# Измененная стоимость продажи
new_sell_price = 3.0

# Новая функция для расчета прибыли с учетом измененной стоимости продажи
def calculate_new_profit_or_loss(purchase_volume):
    profits = []
    for demand, probability in zip(purchase_volumes, probabilities):
        sold_amount = min(purchase_volume, demand)
        unsold_amount = purchase_volume - sold_amount
        profit = (sold_amount * new_sell_price + unsold_amount * discount_price) - purchase_volume * buy_price
        profits.append(profit * probability)
    return sum(profits)

# Расчет ожидаемой прибыли или убытков для каждого объема закупок с новой ценой продажи
new_expected_profits = {volume: calculate_new_profit_or_loss(volume) for volume in purchase_volumes}

# Расчет нового оптимального объема закупок
new_optimal_purchase_volume = max(new_expected_profits, key=new_expected_profits.get)

# Расчет нового стандартного отклонения доходности
def calculate_new_std(purchase_volume):
    profit_variance = []
    expected_profit = calculate_new_profit_or_loss(purchase_volume)
    for demand, probability in zip(purchase_volumes, probabilities):
        sold_amount = min(purchase_volume, demand)
        unsold_amount = purchase_volume - sold_amount
        profit = (sold_amount * new_sell_price + unsold_amount * discount_price) - purchase_volume * buy_price
        profit_variance.append(probability * (profit - expected_profit) ** 2)
    return np.sqrt(sum(profit_variance))

new_std_profits = {volume: calculate_new_std(volume) for volume in purchase_volumes}

# Вычисление новых ожидаемых размеров упущенной выгоды
new_opportunity_losses = {volume: calculate_opportunity_loss(volume) for volume in purchase_volumes}

# Нахождение новых коэффициентов вариации доходности
new_coefficients_of_variation = {volume: calculate_coefficient_of_variation(volume) for volume in purchase_volumes}

# Вычисление нового отношения «доходность/риск»
new_profit_risk_ratio = {volume: new_expected_profits[volume] / new_std_profits[volume] if new_std_profits[volume] != 0 else float('inf') for volume in purchase_volumes}

# Выбор нового оптимального размера закупок, руководствуясь ответами на вопросы 2 и 4
new_optimal_volume_by_profit_and_loss = min(new_opportunity_losses, key=new_opportunity_losses.get)

new_expected_profits, new_optimal_purchase_volume, new_std_profits, new_opportunity_losses, new_coefficients_of_variation, new_profit_risk_ratio, new_optimal_volume_by_profit_and_loss

({500: 750.0, 1000: 1250.0, 2000: 1250.0},
 1000,
 {500: 0.0, 1000: 500.0, 2000: 1500.0},
 {500: 300.0, 1000: 0.0, 2000: 200.0},
 {500: 0.0, 1000: 0.5, 2000: 2.0},
 {500: inf, 1000: 2.5, 2000: 0.8333333333333334},
 1000)

1.  **Ожидаемая прибыль или убытки для каждого объема закупок**:
    
    *   При закупке 500 фунтов: ожидаемая прибыль составляет $750.
    *   При закупке 1000 фунтов: ожидаемая прибыль составляет $1250.
    *   При закупке 2000 фунтов: ожидаемая прибыль составляет $1250.
2.  **Оптимальный объем закупок моллюсков**:
    
    *   Оптимальным объемом закупок по-прежнему является 1000 фунтов, так как ожидаемая прибыль максимальна при этом объеме.
3.  **Стандартное отклонение доходности при закупке**:
    
    *   При закупке 500 фунтов: стандартное отклонение составляет $0.
    *   При закупке 1000 фунтов: стандартное отклонение составляет $500.
    *   При закупке 2000 фунтов: стандартное отклонение составляет $1500.
4.  **Ожидаемые размеры упущенной выгоды**:
    
    *   Остаются неизменными, так как основаны на сравнении с максимальной ожидаемой прибылью.
5.  **Коэффициент вариации доходности**:
    
    *   Остается неизменным, так как зависит от стандартного отклонения и средней прибыли.
6.  **Отношение «доходность/риск»**:
    
    *   При закупке 500 фунтов: бесконечность (нет риска).
    *   При закупке 1000 фунтов: 2.5.
    *   При закупке 2000 фунтов: 0.83.
7.  **Оптимальный размер закупок, руководствуясь ответами на вопросы 2 и 4**:
    
    *   Оптимальный объем закупок остается 1000 фунтов.

In [5]:
# Пункт 10 - Изменение вероятностей заказов на 0.4, 0.4 и 0.2 для объемов 500, 1000 и 2000 фунтов соответственно

# Измененные вероятности
new_probabilities = [0.4, 0.4, 0.2]

# Пересчет ожидаемой прибыли или убытков с новыми вероятностями
def calculate_profit_or_loss_with_new_probabilities(purchase_volume):
    profits = []
    for demand, probability in zip(purchase_volumes, new_probabilities):
        sold_amount = min(purchase_volume, demand)
        unsold_amount = purchase_volume - sold_amount
        profit = (sold_amount * sell_price + unsold_amount * discount_price) - purchase_volume * buy_price
        profits.append(profit * probability)
    return sum(profits)

new_expected_profits_with_probabilities = {volume: calculate_profit_or_loss_with_new_probabilities(volume) for volume in purchase_volumes}

# Расчет нового оптимального объема закупок с измененными вероятностями
new_optimal_purchase_volume_with_probabilities = max(new_expected_profits_with_probabilities, key=new_expected_profits_with_probabilities.get)

# Расчет нового стандартного отклонения доходности с измененными вероятностями
def calculate_new_std_with_probabilities(purchase_volume):
    profit_variance = []
    expected_profit = calculate_profit_or_loss_with_new_probabilities(purchase_volume)
    for demand, probability in zip(purchase_volumes, new_probabilities):
        sold_amount = min(purchase_volume, demand)
        unsold_amount = purchase_volume - sold_amount
        profit = (sold_amount * sell_price + unsold_amount * discount_price) - purchase_volume * buy_price
        profit_variance.append(probability * (profit - expected_profit) ** 2)
    return np.sqrt(sum(profit_variance))

new_std_profits_with_probabilities = {volume: calculate_new_std_with_probabilities(volume) for volume in purchase_volumes}

new_expected_profits_with_probabilities, new_optimal_purchase_volume_with_probabilities, new_std_profits_with_probabilities

({500: 500.0, 1000: 600.0, 2000: 0.0},
 1000,
 {500: 0.0, 1000: 489.89794855663564, 2000: 1095.4451150103323})

Результаты с измененными вероятностями заказов (0.4, 0.4 и 0.2 для объемов 500, 1000 и 2000 фунтов соответственно)

1.  **Ожидаемая прибыль или убытки для каждого объема закупок**:
    
    *   При закупке 500 фунтов: ожидаемая прибыль составляет $500.
    *   При закупке 1000 фунтов: ожидаемая прибыль составляет $600.
    *   При закупке 2000 фунтов: ожидаемая прибыль равна $0.
2.  **Оптимальный объем закупок моллюсков**:
    
    *   Оптимальным объемом закупок в данной ситуации остается 1000 фунтов, так как при этом объеме ожидаемая прибыль максимальна.
3.  **Стандартное отклонение доходности при закупке**:
    
    *   При закупке 500 фунтов: стандартное отклонение составляет $0.
    *   При закупке 1000 фунтов: стандартное отклонение составляет примерно $489.90.
    *   При закупке 2000 фунтов: стандартное отклонение составляет примерно $1095.45.

Эти результаты показывают, как изменение вероятностей спроса влияет на ожидаемую прибыль и оптимальный объем закупок. С изменением вероятностей, максимальная ожидаемая прибыль по-прежнему достигается при закупке 1000 фунтов, хотя общий уровень прибыли снизился по сравнению с первоначальными условиями.

# Часть 2

In [6]:
# Данные из задания
sales_volumes = [1000, 2000, 5000, 10000, 50000]  # Варианты тиража
probabilities = [0.45, 0.20, 0.15, 0.1, 0.1]  # Вероятности соответствующих заказов

# Условия контракта с издательствами
company_A = {'fixed': 10000, 'per_book': 2}
company_B = {'fixed': 2000, 'per_book': 4}

# Функция для расчета ожидаемой прибыли от контракта с издательством
def calculate_expected_profit(contract, sales_volumes, probabilities):
    expected_profits = []
    for sales_volume, probability in zip(sales_volumes, probabilities):
        profit = contract['fixed'] + contract['per_book'] * sales_volume
        expected_profits.append(profit * probability)
    return sum(expected_profits)

# Вычисление ожидаемой прибыли
expected_profit_A = calculate_expected_profit(company_A, sales_volumes, probabilities)
expected_profit_B = calculate_expected_profit(company_B, sales_volumes, probabilities)

# Вычисление стандартного отклонения для оценки риска
def calculate_std_deviation(contract, sales_volumes, probabilities, expected_profit):
    variances = []
    for sales_volume, probability in zip(sales_volumes, probabilities):
        profit = contract['fixed'] + contract['per_book'] * sales_volume
        variance = (profit - expected_profit) ** 2
        variances.append(variance * probability)
    return np.sqrt(sum(variances))

std_deviation_A = calculate_std_deviation(company_A, sales_volumes, probabilities, expected_profit_A)
std_deviation_B = calculate_std_deviation(company_B, sales_volumes, probabilities, expected_profit_B)

# Вычисление коэффициентов вариации и отношений «доходность/риск»
coefficient_of_variation_A = std_deviation_A / expected_profit_A
coefficient_of_variation_B = std_deviation_B / expected_profit_B
profit_risk_ratio_A = expected_profit_A / std_deviation_A if std_deviation_A != 0 else float('inf')
profit_risk_ratio_B = expected_profit_B / std_deviation_B if std_deviation_B != 0 else float('inf')

expected_profit_A, expected_profit_B, std_deviation_A, std_deviation_B, coefficient_of_variation_A, coefficient_of_variation_B, profit_risk_ratio_A, profit_risk_ratio_B

(25200.0,
 32400.0,
 28791.665460684973,
 57583.33092136995,
 1.1425264071700387,
 1.7772633000422824,
 0.8752532928117898,
 0.5626628310932934)

Вот результаты анализа для задачи:

1.  **Ожидаемая прибыль от контракта с издательствами**:
    
    *   С издательством А: $25,200.
    *   С издательством Б: $32,400.
2.  **Ожидаемые размеры упущенной выгоды**:
    
    *   Упущенная выгода - это разница в прибыли между двумя вариантами. Таким образом, упущенная выгода от контракта с издательством А составляет $7,200 (32,400 - 25,200), а с издательством Б - $0, так как это более выгодный вариант.
3.  **Выбор издательства для заключения договора**:
    
    *   Руководствуясь ответами на вопросы 1 и 2, следует заключить договор с издательством Б, так как ожидаемая прибыль выше и упущенная выгода меньше.
4.  **Коэффициент вариации прибыли от контрактов**:
    
    *   С издательством А: 1.14.
    *   С издательством Б: 1.78.
5.  **Отношение «доходность/риск»**:
    
    *   С издательством А: 0.88.
    *   С издательством Б: 0.56.
6.  **Принятие оптимального решения**:
    
    *   Руководствуясь коэффициентом вариации и отношением «доходность/риск», издательство А представляет собой более стабильный и менее рискованный вариант. Несмотря на меньшую ожидаемую прибыль, издательство А имеет более высокое отношение «доходность/риск» и меньший коэффициент вариации, что указывает на меньшую волатильность доходов.
7.  **Сравнение результатов решения задач 3 и 6**:
    
    *   В третьем пункте было выбрано издательство Б на основе максимальной ожидаемой прибыли. Однако в шестом пункте, учитывая риск и стабильность доходов, предпочтительнее оказалось издательство А. Это показывает, что выбор между издательствами зависит не только от ожидаемой прибыли, но и от уровня риска. Хотя издательство Б обещает более высокую прибыль, издательство А предлагает большую стабильность и меньший риск, что может быть предпочтительнее для некоторых авторов, особенно тех, кто ценит надежность и предсказуемость дохода.

    Этот анализ показывает, что выбор между издательствами А и Б зависит от того, насколько автор готов рисковать для получения более высокой прибыли, и от его предпочтений в отношении стабильности дохода.

In [7]:
# Измененные условия контракта с издательством А
company_A_updated = {'fixed': 10000, 'per_book': 3}

# Пересчет ожидаемой прибыли с измененными условиями контракта А
expected_profit_A_updated = calculate_expected_profit(company_A_updated, sales_volumes, probabilities)

# Пересчет стандартного отклонения для оценки риска с измененными условиями контракта А
std_deviation_A_updated = calculate_std_deviation(company_A_updated, sales_volumes, probabilities, expected_profit_A_updated)

# Вычисление коэффициентов вариации и отношений «доходность/риск» с измененными условиями контракта А
coefficient_of_variation_A_updated = std_deviation_A_updated / expected_profit_A_updated
profit_risk_ratio_A_updated = expected_profit_A_updated / std_deviation_A_updated if std_deviation_A_updated != 0 else float('inf')

# Ожидаемые размеры упущенной выгоды с измененными условиями контракта А
opportunity_loss_A_updated = expected_profit_B - expected_profit_A_updated
opportunity_loss_B_updated = 0  # Издательство Б по-прежнему предлагает большую прибыль

expected_profit_A_updated, std_deviation_A_updated, coefficient_of_variation_A_updated, profit_risk_ratio_A_updated, opportunity_loss_A_updated, opportunity_loss_B_updated

(32800.0, 43187.49819102746, 1.3166920180191297, 0.7594790477308653, -400.0, 0)

С измененными условиями контракта от издательства А (3 доллара за каждую проданную книгу), вот результаты анализа:

1.  **Ожидаемая прибыль от контракта с издательствами**:
    
    *   С издательством А (после изменения): $32,800.
    *   С издательством Б: $32,400 (остается неизменной).
2.  **Ожидаемые размеры упущенной выгоды**:
    
    *   С издательством А (после изменения): упущенная выгода составляет -$400 (на самом деле прибыль выше, чем у издательства Б).
    *   С издательством Б: упущенная выгода равна $0.
3.  **Выбор издательства для заключения договора**:
    
    *   После изменения условий контракта с издательством А, теперь ожидаемая прибыль от издательства А выше, чем от издательства Б. Таким образом, выбор издательства А становится предпочтительнее.
4.  **Коэффициент вариации прибыли от контрактов**:
    
    *   С издательством А (после изменения): 1.32.
    *   С издательством Б: 1.78 (остается неизменным).
5.  **Отношение «доходность/риск»**:
    
    *   С издательством А (после изменения): 0.76.
    *   С издательством Б: 0.56 (остается неизменным).
6.   **Принятие оптимального решения**:
    
    *   Руководствуясь коэффициентом вариации и отношением «доходность/риск», издательство А (после изменения) теперь представляется как более предпочтительный вариант, несмотря на несколько более высокий коэффициент вариации по сравнению с первоначальным вариантом. Оно обеспечивает высокую прибыль при умеренном уровне риска.
7.   **Сравнение результатов решения задач 3 и 6**:
    
    *   Изменение условий контракта с издательством А привело к тому, что оно стало более привлекательным с точки зрения ожидаемой прибыли, упущенной выгоды и отношения «доходность/риск». Это демонстрирует, как изменения в условиях договора могут повлиять на решение автора, подчеркивая важность тщательного анализа всех аспектов предложения при принятии решений.

In [8]:
new_probabilities = [0.3, 0.2, 0.2, 0.1, 0.2]

# Пересчет ожидаемой прибыли с измененными вероятностями для обоих издательств
expected_profit_A_new_prob = calculate_expected_profit(company_A_updated, sales_volumes, new_probabilities)
expected_profit_B_new_prob = calculate_expected_profit(company_B, sales_volumes, new_probabilities)

# Пересчет стандартного отклонения для обоих издательств с измененными вероятностями
std_deviation_A_new_prob = calculate_std_deviation(company_A_updated, sales_volumes, new_probabilities, expected_profit_A_new_prob)
std_deviation_B_new_prob = calculate_std_deviation(company_B, sales_volumes, new_probabilities, expected_profit_B_new_prob)

# Вычисление коэффициентов вариации и отношений «доходность/риск» для обоих издательств с измененными вероятностями
coefficient_of_variation_A_new_prob = std_deviation_A_new_prob / expected_profit_A_new_prob
coefficient_of_variation_B_new_prob = std_deviation_B_new_prob / expected_profit_B_new_prob
profit_risk_ratio_A_new_prob = expected_profit_A_new_prob / std_deviation_A_new_prob if std_deviation_A_new_prob != 0 else float('inf')
profit_risk_ratio_B_new_prob = expected_profit_B_new_prob / std_deviation_B_new_prob if std_deviation_B_new_prob != 0 else float('inf')

# Ожидаемые размеры упущенной выгоды с измененными вероятностями для обоих издательств
opportunity_loss_A_new_prob = expected_profit_B_new_prob - expected_profit_A_new_prob
opportunity_loss_B_new_prob = 0  # Издательство Б по-прежнему предлагает большую прибыль

expected_profit_A_new_prob, expected_profit_B_new_prob, std_deviation_A_new_prob, std_deviation_B_new_prob, coefficient_of_variation_A_new_prob, coefficient_of_variation_B_new_prob, profit_risk_ratio_A_new_prob, profit_risk_ratio_B_new_prob, opportunity_loss_A_new_prob, opportunity_loss_B_new_prob

(48100.0,
 52800.0,
 56509.20279034203,
 75345.60372045604,
 1.1748275008387117,
 1.4270000704631827,
 0.8511887909383276,
 0.7007708133296834,
 4700.0,
 0)

Результаты анализа с измененными вероятностями заказов (0.3, 0.2, 0.2, 0.1, 0.2),

1.  **Ожидаемая прибыль от контрактов**:
    
    *   С издательством А (после изменения условий): ожидаемая прибыль составляет $48,100.
    *   С издательством Б: ожидаемая прибыль составляет $52,800.
2.  **Ожидаемые размеры упущенной выгоды**:
    
    *   С издательством А (после изменения условий): упущенная выгода составляет $4,700.
    *   С издательством Б: упущенная выгода отсутствует, так как это более выгодный вариант.
3.  **Выбор издательства для заключения договора**:
    
    *   С учетом новых вероятностей, издательство Б предлагает более высокую ожидаемую прибыль, что делает его предпочтительным выбором.
4.  **Коэффициент вариации прибыли от контрактов**:
    
    *   С издательством А (после изменения): 1.17.
    *   С издательством Б: 1.43.
5.  **Отношение «доходность/риск»**:
    
    *   С издательством А (после изменения): 0.85.
    *   С издательством Б: 0.70.
6.  **Принятие оптимального решения**:
    
    *   На основе коэффициента вариации и отношения «доходность/риск», издательство А представляется как менее рискованный вариант. Несмотря на то, что ожидаемая прибыль от издательства Б выше, издательство А предлагает более стабильный доход с меньшим риском.
7.   **Сравнение результатов решения задач 3 и 6**:
    
    *   С учетом измененных вероятностей, издательство Б остается более выгодным вариантом с точки зрения максимальной ожидаемой прибыли. Однако, если учитывать риск и стабильность доходов, издательство А может быть предпочтительным из-за более низкого коэффициента вариации и высокого отношения «доходность/риск».
    
С изменением вероятностей, общая ожидаемая прибыль у обоих издательств увеличилась, однако издательство Б продолжает предлагать более высокую прибыль, хотя с более высоким уровнем риска. Это подчеркивает важность оценки вероятностных распределений при принятии решений, связанных с доходностью и риском.
